File: core/api/typescript/src/lib/services/shared_path.ts
import type { GenericPathJoiner } from "../enclaves/generic_path_joiner";

// Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
export class SharedPath {

    //Absolute path in the container where this code is running
    private readonly absPathOnThisContainer: string
    //Absolute path in the service container
    private readonly absPathOnServiceContainer: string
    private readonly pathJoiner: GenericPathJoiner;

    constructor (absPathOnThisContainer: string, absPathOnServiceContainer: string, pathJoiner: GenericPathJoiner) {
        this.absPathOnThisContainer = absPathOnThisContainer;
        this.absPathOnServiceContainer = absPathOnServiceContainer;
        this.pathJoiner = pathJoiner
    }
    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getAbsPathOnThisContainer(): string {
        return this.absPathOnThisContainer;
    }
    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getAbsPathOnServiceContainer(): string {
        return this.absPathOnServiceContainer;
    }
    // Docs available at https://docs.kurtosistech.com/kurtosis-core/lib-documentation
    public getChildPath(pathElement: string): SharedPath {
        const absPathOnThisContainer = this.pathJoiner.join(this.absPathOnThisContainer, pathElement);
        const absPathOnServiceContainer = this.pathJoiner.join(this.absPathOnServiceContainer, pathElement);

        const sharedPath = new SharedPath(absPathOnThisContainer, absPathOnServiceContainer, this.pathJoiner)
        return sharedPath
    }
}

File: core/api/golang/lib/services/shared_path.go
package services

import (
	"path/filepath"
)

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
type SharedPath struct {
	//Absolute path in the container where this code is running
	absPathOnThisContainer string
	//Absolute path in the service container
	absPathOnServiceContainer string
}

func NewSharedPath(absPathOnThisContainer string, absPathOnServiceContainer string) *SharedPath {
	return &SharedPath{absPathOnThisContainer: absPathOnThisContainer, absPathOnServiceContainer: absPathOnServiceContainer}
}
// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (s SharedPath) GetAbsPathOnThisContainer() string {
	return s.absPathOnThisContainer
}
// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (s SharedPath) GetAbsPathOnServiceContainer() string {
	return s.absPathOnServiceContainer
}

// Docs available at https://docs.kurtosistech.com/kurtosis-client/lib-documentation
func (s SharedPath) GetChildPath(pathElement string) *SharedPath {

	absPathOnThisContainer := filepath.Join(s.absPathOnThisContainer, pathElement)

	absPathOnServiceContainer := filepath.Join(s.absPathOnServiceContainer, pathElement)

	sharedPath := NewSharedPath(absPathOnThisContainer, absPathOnServiceContainer)

	return sharedPath
}
